### PR TITLE
[WIP] RL for Math

### PR DESCRIPTION
**Pull request by AlexPiche**
# RL loop

```python
python examples/rl_gsm8k/orchestrate_rl.py
```

## Overview

image

**collect data**
* the current model (updated llama 3.1 8b) is served on all the gpus using vllm. 
* a subset of 16 tasks from the train set of gsm8k is sampled and replicated 64 times each for a total of 1024 tasks. 
* the agent produce complete tapes for each of these 1024 tasks using temperature 0.7. 
* traces are created from these new tapes. 
* the log prob of the traces under the current model are computed. I refer to these as old log prob to be close to the naming of the grpo paper. #todo should be log prob and old log prob only in fine tune. 
* the current model is taken down and the reference model is now served on all gpus using vllm. 
* the log prob from the reference model (llama 3.1 8b) are computed for the most recent traces. 

**Evaluation**
* every 5 iterations, the agent is evaluated with temperature 0 on the complete test set of gsm8k. No traces are produced on the test set. 

**Finetuning**
* RL (grpo or reinforce) training is performed on the latest batch of data using a separate process. This makes it easier to manage memory with the main process. Otherwise there are sometimes issues with the gpus not being completely empty after fine tuning. 

## Fix metadata instantation bug

A bug was identified in the metadata creation process, resulting in all math tapes having the same ID. This was caused by an incorrect implementation of the TapeMetadata creation. The issue was resolved by changing the way TapeMetadata is instantiated. 

**Before:** metadata was initialized directly with `TapeMetadata()`, which could lead to all instances referencing the same metadata object.

**After:** metadata is now initialized with `Field(default_factory=TapeMetadata)`, ensuring that each instance of the containing class gets a unique `TapeMetadata` object.

## Batch main loop

* Takes `max_loop` parameter with default -1 to keep previous behavior.
* Return tape now have the `parent_id` set to the original tape id

## Upgrade transformers and tokenizers

The fine tune script depends on correct spans for the tokenizer and the llama 3 tokenizer has been broken for several months. Only on the latest version of `tokenizers` was this issue fixed. I upgraded `tokenizers` and had to also upgrade `transformers` for compatibility. 

## SQLite queue

When used with `batch_main_loop`, the sql database locks delaying the loop (the gpus are then sitting idle). I've implemented a queue such that sql operations are done sequentially instead of in parallel which prevent locking of the database. I've benchmarked the change and it results in a little less than 2x throughput. # todo: I implemented a naive version of this and want to consult Oleh about the best way to implement this correctly in the code base. 

### Queue benchmark
Setup
* 8 H100s
* Llama 3.1 8B
* Collecting 1024 tapes
* max loop 1
* 1 replication

#### Without queue

| Workers | Tapes per second |
|---------|------------------|
| 64      | 3.58             |
| 32      | 3.29             |
| 128     | N/A (Failed to store LLMCall: database is locked) |

#### Using queue

| Workers | Tapes per second |
|---------|------------------|
| 128     | 5.91             |
| 256     | 5.85             |
| 64      | 6.25             |
| 32      | 6.46             |
| 16      | 5.46             |

## Fix training sample creation

The training sample creation (`make_training_text`) did not correctly computed the number of predicted characters (`n_predicted`) and included part of the prompt as predicted outputs. 

**End of prompt before**

Screenshot 2024-10-24 at 3 04 53 PM

**End of prompt after**

Screenshot 2024-10-24 at 3 04 58 PM

## Add fields to training samples

I added the reward, logprobs, ref logprobs, and parent tape id fields to the training samples class. 

Screenshot 2024-10-28 at 11 04 28 AM

## fix tests

Adding a new field to training samples resulted in a few tests failing. I've updated these tests to handle the new fields. 

## get log probs 

The log probs of a completion must often be computed in RL. I implemented log probs computation both for chat complete and complete. I validate that the returned log prob tokens match the HF tokens. 

## RL finetune

Most RL algorithms share roughly the same pipeline with the exception of the loss. The user can now switch between 2 RL losses in the config by changing `finetune.rl.algo` to either `reinforce` or `grpo`.

## Performance

Screenshot 2024-10-25 at 2 10 54 PM

**Note** Performance becomes unstable later during training for GRPO. I suggest to merge this branch and open an issue to fix it in a later PR.

## Execution time

**Setup**

* Llama 3.1 8b
* 8 H100s
* 1024 tapes 

**RL loop**

Screenshot 2024-10-25 at 9 23 47 AM

* Most of the time is spent fine tuning
* A considerable amount of time is spent on populating ref log probs.

**Zooming on make data**

Screenshot 2024-10-25 at 9 21 25 AM

* LLM inference is half of the time spent on collecting new data
* Annotate tapes consist of:
  * Getting the log probs of the completion (most of the time)
  * Assigning rewards
* Reading from the sqlite dataset is negligible, since we keep the dataset small by discarding the completions from previous iterations.

****

## Checklist

- [x] serve local model with vllm 
- [x] Get log probs for both complete and chat_complete 
- [x] Fix make prompt
- [x] Update tokenizers and transformers (fix the llama3 tokenizer issue)
- [x] Iterative RL
- [x] Multi gpu RL inference and training
- [x] Create metadata everytime the class MathTape is called and not only once (was creating a single id for multiple tapes)
- [x] Add queue to write to SQLite 
- [x] Better training text field names
- [x] Reward go up

### CODE REVIEW

**Comment by rizar** [comment]
I haven't looked at this yet, but it might  speed up the merge process if you (a) resolve merge conflicts (b) write a readme.

File: tapeagents/finetune/rl/__init__.py [hunk 1]
```diff
@@ -119,69 +122,60 @@ def grpo_step(model, batch, config: GRPOConfig) -> tuple[torch.Tensor, dict[str,
     masks_ = masks[:, 1:]
     ref_logprobs = batch["ref_logprobs"][:, 1:]
     old_logprobs = batch["old_logprobs"][:, 1:]
```
Comment (rafapi) [inline tapeagents/finetune/rl/__init__.py:124]: do we have old_logprobs anymore?

File: tapeagents/finetune/rl/__init__.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 logger = logging.getLogger(__name__)
 
 RL_DATA_COLUMNS = [
+    "reward",
     "rewards",
     "advantages",
     "old_logprobs",
```
Comment (rafapi) [inline tapeagents/finetune/rl/__init__.py:24]: ref_logprobs instead of old_logprobs?

File: tapeagents/finetune/rl/__init__.py [hunk 3]
```diff
@@ -119,69 +122,60 @@ def grpo_step(model, batch, config: GRPOConfig) -> tuple[torch.Tensor, dict[str,
...
-    # assert torch.isfinite(advantages).all(), "advantages contains NaN or inf"
-    # assert torch.isfinite(ref_logprobs).all(), "ref_logprobs contains NaN or inf"
-    # assert torch.isfinite(old_logprobs).all(), "old_logprobs contains NaN or inf"
-    # assert torch.isfinite(new_log_probs).all(), "new_log_probs contains NaN or inf"
-
     assert new_log_probs.shape == ref_logprobs.shape
 
     # First compute the PPO surrogate loss, see https://arxiv.org/pdf/2402.03300 eq 3
     log_ratio_new_old = new_log_probs - old_logprobs
     log_ratio_new_old = new_log_probs - old_logprobs
```
Comment (rafapi) [inline tapeagents/finetune/rl/__init__.py:129]: duplicated line

File: tapeagents/finetune/rl/__init__.py (hunk 1 above)
Comment (AlexPiche) [reply tapeagents/finetune/rl/__init__.py:124]: we do

File: tapeagents/core.py [hunk 4]
```diff
@@ -13,10 +13,21 @@
 class TrainingText(BaseModel):
     """
     Data sample to finetune a language model
+    A training sample for finetuning a language model. Contains:
+    - text: The full text sequence (prompt + completion)
+    - n_predicted: Number of tokens in the completion
+    - reward: Reward value for the completion
```
Comment (rizar) [inline tapeagents/core.py:19]: can we use "output" instead of "completion" to adhere to `LLM` interface in TapeAgents? 

instead of "assistant model" I'd use "current model"

and maybe can we separate the fields that are necessary for RL from other fields

Comment (rizar) [reply tapeagents/core.py:19]: also I'd suggest `parent_tape_id` -> `tape_id`, the shorter name is sufficiently descriptive here

File: tapeagents/core.py [hunk 5]
```diff
@@ -144,9 +157,9 @@ class Tape(BaseModel, Generic[ContextType, StepType]):
     A sequence of steps produced by agents and environments
     """
 
-    metadata: TapeMetadata = TapeMetadata()
+    metadata: TapeMetadata = Field(default_factory=TapeMetadata)
```
Comment (rizar) [inline tapeagents/core.py:160]: great fix!

File: tapeagents/finetune/checkpoints.py [hunk 6]
```diff
@@ -35,7 +35,7 @@ def get_auto_model_class(
 
 
 def load_tokenizer(config_name):
-    tokenizer = AutoTokenizer.from_pretrained(config_name, use_fast=True)
+    tokenizer = AutoTokenizer.from_pretrained(config_name, use_fast=True, truncation_side="left")
```
Comment (rizar) [inline tapeagents/finetune/checkpoints.py:38]: @ollmer , can you please review this change? I'm not up-to-speed here

File: examples/rl_gsm8k/orchestrate_rl.py [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+    os.makedirs(conf_dir, exist_ok=True)
+    finetune_path = exp_path / "finetune"
+    while state["iteration"] <= cfg.max_iterations:
+        if os.path.exists(finetune_path / "current"):
+            assistant_model_path = str(finetune_path / "current")
+        else:
+            assistant_model_path = cfg.model_path
+
+        start_serving_assistant = time.time()
+        assistant_process, stdout_file, stderr_file = serve_vllm_local_model(
```
Comment (rizar) [inline examples/rl_gsm8k/orchestrate_rl.py:436]: one day we should make this a context manager

**Review by rizar — CHANGES_REQUESTED** [review]
This is a really, really great contribution! 

Few things we need to improve or discuss before the merge:
- I think some changes in `agent.py` might not be necessary
- I'm afraid that the SQLite queue is opening up a dangerous can of worms, see my comment below about what are the guarantees for having all LLMCalls dumped to the db
- I have not found README.md

File: tapeagents/agent.py [hunk 8]
```diff
@@ -283,9 +290,11 @@ def delegate(self, tape: TapeType) -> Agent[TapeType]:
             subagent = subagent.find_subagent(view.agent_name)
         return subagent
 
-    def is_agent_step(self, step: Step) -> bool:
-        """Check if the step was produced by the agent or by the environment."""
-        return isinstance(step, (Action, Thought))
+    def is_llm_step(self, step: Step) -> bool:
```
[the remark this answers is not in the document]
Comment (rizar) [reply tapeagents/agent.py:293]: discussed with @ollmer , we should revert this change

File: tapeagents/agent.py [hunk 9]
```diff
@@ -373,7 +382,7 @@ def reuse(self, tape: TapeType) -> tuple[TapeType, list[LLMCall]]:
         while i < len(tape):
             past_tape = tape[:i]
             step = tape.steps[i]
-            if self.is_agent_step(step):
+            if self.is_llm_step(step):
```
[the remark this answers is not in the document]
Comment (rizar) [reply tapeagents/agent.py:385]: discussed with @ollmer , we should revert this change

**Review by rizar — APPROVED** [review]
great addition!

File: examples/gsm8k_tuning/math_agent.py [hunk 10]
```diff
@@ -190,7 +191,7 @@ def solve_task(agent: Agent, env: Environment, task: dict, tape_file: str = "")
 
 def extract_result_value(sample) -> dict:
     sample = dict(sample)
-    expected_result = str(sample["answer"]).rsplit("####", maxsplit=1)[-1].strip().replace(",", ".")
+    expected_result = str(sample["answer"]).rsplit("####", maxsplit=1)[-1].strip().replace(",", "")
```
Comment (ollmer) [inline examples/gsm8k_tuning/math_agent.py:194]: It should break floats written as string, no?

Comment (AlexPiche) [reply examples/gsm8k_tuning/math_agent.py:194]: Hi @ollmer , just saw this comment now. Can we chat about this in Monday? With the previous implementation, I was getting the following error: sample["answer"] = "1,000,000", was getting converted to "1.000.000" which could not be parsed as a int.

Comment (ollmer) [reply examples/gsm8k_tuning/math_agent.py:194]: Sure. Also, let's check the dataset for decimal separator consistency. Maybe they're a bit different between samples
